text/0000-pi-types.md

```diff
@@ -0,0 +1,644 @@
+(`const fn`) to the parameter, to perform compile time, type level calculations
+on the parameter. This allows for great expressiveness as `const fn` improves.
+
+## Expression `where` bounds
+
+The second construct added is the constant expression in `where` bounds. These
+contains statements about the constant parameters, which are checked at compile
+time.
+
+## Type checking
+
+Type checking is done by using a [transitive rule](#transitivity-of-bounds), such that `where` bounds must
+be implied by the caller.
+
+# Motivation
+[motivation]: #motivation
+
+An often requested feature is the "type-level numerals", which enables generic
+length arrays. The current proposals are often limited to integers or even lack
+of value maps, and other critical features.
+
+There is a whole lot of other usecases as well. These allows certain often
+requested features to live in standalone libraries (e.g., [bounded-integers](#bounded-integersinterval-arithmetics),
+[type level numerals](#array-generics), [statically checked indexing](#statically-checked-indexing), lattice types).
+
+It allows for creating powerful abstractions without type-level hackery.
+
+## What we want, and what we don't want
+
+We have to be very careful to avoid certain things, while still preserving the core features:
+
+1. Ability to use, manipulate, and constraint values at type-level.
+2. The ability to use said values on expression-level (runtime).
+
+Yet, we do not want:
+
+1. SMT-solvers, due to not only undecidability (note, although, that SAT is
+   decidable) and performance, but the complications it adds to `rustc`.
+2. Monomorphisation-time errors, i.e. errors that happens during codegen of
+   generic functions. We try to avoid adding _more_ of these (as noted by
+   petrochenkov, these [already exists](https://github.com/rust-lang/rfcs/pull/1657#discussion_r68202733))
+
+## Examples
+
+### Bounded integers/interval arithmetics
+
+One can define the so called "bounded integers" (integers which carry an
+upper and lower bound, checked at compile time):
+
+```rust
+use std::ops;
+
+/// A bounded integer.
+///
+/// This has two value parameter, respectively representing an upper and a lower bound.
+pub struct BoundedInt<const lower: usize, const upper: usize>
+    where lower <= upper {
+    /// The inner runtime value.
+    n: usize,
+}
+
+// To see how this holds the `where` clause above, see the section on `identities`.
+impl<const n: usize> BoundedInt<n, n> {
+    fn new() -> Self {
+        BoundedInt {
+            n: n,
+        }
+    }
+}
+
+/// Addition of two `BoundedInt` will simply add their bounds.
+///
+/// We check for overflow making it statically overflow-free calculations.
+impl<const upper_a: usize,
+     const lower_a: usize,
+     const upper_b: usize,
+     const lower_b: usize> ops::Add<BoundedInt<lower_b, upper_b>> for BoundedInt<lower_a, lower_b>
+     where lower_a <= upper_a,
+           lower_b <= upper_b,
```

Comment:
(add comment about why these are reuquired and why they arent implied by the `BoundedInt` where clause like in the `impl new` example above.